### PR TITLE
feat: implement proper error handling for requester

### PR DESCRIPTION
**Pull request by GregoryAlbouy**
## Description

This PR implements proper error handling for `requester` and `dispatcher` (formerly `semimpl`).

### Refactoring: `semimpl` ➡️ `dispatcher`
  ```go
  // before
  semimpl.Do(ctx, 100, 10, callback) // notice the error?
  
  // after
  dispatcher.New(10).Do(ctx, 100, callback)
  ```

### Error handling: `dispatcher`

- `dispatcher.New` **panics** if `numWorker < 1`
- `dispatcher.Dispatcher.Do` returns a non-nil error early if
  - `maxIter < 1 && maxIter != -1`
  - `maxIter < Dispatcher.numWorker && maxIter != -1`
  - `callback == nil`
- New unit tests for all cases

### Error handling: `requester`
- Handle `config.HTTPRequest()` error in `Dispatcher.Run` instead of repeated callback
- Handle `client.Do` connection error in `Dispatcher.Run` in addition to the callback via a ping to return early in case of bad URL, server not started, ...

### CODE REVIEW

**Review by c1-ra — DISMISSED** [review]
[question] Why do you think this way is less prone to error?
I am ok you merge it anyway, I have really no preference between this implementation and the previous one, both seem ok to me!

**Comment by GregoryAlbouy** [comment]
It is less error prone because it avoids this kind of error:
```go
semimpl.Do(ctx, 100, 10, callback) // notice the error?
                                   // numWorker and maxIter are inverted: we have more workers than iterations
```

File: dispatcher/dispatcher.go
Comment (moreirathomas) [inline dispatcher/dispatcher.go:47]: At this point why not inline this function ?
```suggestion
	if numWorkers < 1 {
		numWorkers = 1
	}
```

Same for the 2 others?

**Review by moreirathomas — COMMENTED** [review]
Seems good.

Same as Clara, I don't know if an other data struct is really necessary but
> ```go
> semimpl.Do(ctx, 100, 10, callback) // notice the error?
>                                    // numWorker and maxIter are inverted: we have more workers than iterations
> ```
is valid so I'll run with it.

Just waiting for reply on the other comment before approving.

File: dispatcher/dispatcher.go
Comment (c1-ra) [reply dispatcher/dispatcher.go:47]: Since we are supposed to call `Validate()` on the config before declaring the dispatcher, couldn't we just get rid of all the sanitize methods because we can suppose that all configurations are valid?

**Comment by GregoryAlbouy** [comment]
>  I don't know if an other data struct is really necessary

Clearly none of the changes in this PR is necessary. Only debatable changes from `semimpl` that I consider an API improvement.

One notably questionable consequence is regarding the error handling, we probably don't want to check both `New` and `Do`:
```go
d, err := dispatcher.New(-4)
if err != nil {}
if err = d.Do(...) {}
```
Because this si clearly not an AP improvement. We'll want to keep it chainable `dispatcher.New(10).Do(...)` so we'll have to silently sanitize `numWorkers` in `New` and do the error handling in `Do` only (which is perfectly ok to me)

File: dispatcher/dispatcher.go
Comment (GregoryAlbouy) [reply dispatcher/dispatcher.go:47]: To me `dispatcher` should remain a completely independent package with its own logic, and making assumptions on its inputs based on how it's used outside is already a form of couplage. The external config validation is subject to change, and this package shouldn't depend on it. For that reason I believe it should have its own error handling.

As for the inlining, agreed!

File: dispatcher/dispatcher_test.go
Comment (c1-ra) [inline dispatcher/dispatcher_test.go:257]: `exp: errors.New("invalid value: maxIter: must be >= 1, got 0")`
I think it should be **>=1**?

File: dispatcher/dispatcher.go
Comment (c1-ra) [inline dispatcher/dispatcher.go:58]: `return fmt.Errorf("%w: maxIter: must be >= 1, got %d", ErrInvalidValue, maxIter)`

**Comment by c1-ra** [comment]
I just added two minor comments, otherwise, it seems good :)

File: dispatcher/dispatcher.go
Comment (GregoryAlbouy) [reply dispatcher/dispatcher.go]: Good catch! Not minor at all, totally forgot that use case. Fixed!

File: requester/requester.go
Comment (moreirathomas) [inline requester/requester.go:34]: Side not but super important: now that `-1` for number of requests is allowed and expected, calling `make(chan Record, cfg.RunnerOptions.Requests)` can panic because `make` do not accept negative integers.

File: requester/requester.go
Comment (moreirathomas) [inline requester/requester.go:60]: Very nice !

File: requester/requester.go
Comment (moreirathomas) [inline requester/requester.go:78]: I might be wrong but:
I'm pretty sure reading from `errCh` will block until something from `requester.Do` writes to it.

Meaning `r.collect()` is not run in parallel of `Do`. Losing the whole point.

---

Suggestion/path to explore, untested:

```suggestion
	return r.collect(ctx, errCh)
```
And this `r.collect`:
```go
func (r *Requester) collect(ctx context.Context, errc <-chan error) (Report, error) {
	rep := Report{
		Config:  r.config,
		Records: make([]Record, 0, r.config.RunnerOptions.Requests), // Provide capacity if known.
	}
	for {
		select {
		case <-ctx.Done():
			return rep, nil
		case err := <-errc:
			return Report{}, err
		case rec := <-r.records:
			if rec.Error != nil {
				rep.Fail++
			} else {
				rep.Success++
			}
			rep.Records = append(rep.Records, rec)
			rep.Length++
		}
	}
}
```

Note: we go back to incremental Report building.

Maybe going back to safe slice and mutex is better.

File: requester/requester.go
Comment (GregoryAlbouy) [reply requester/requester.go:34]: Indeed, noticed the same for `collect`:
https://github.com/benchttp/runner/blob/300bfb6438467cfd86ea38bf58871a6e423e7a6f/requester/report.go#L30-L34

Hence the addition in the docs 3 lines above this comment:
```go
// New returns a Requester initialized with cfg. cfg is assumed valid:
// it is the caller's responsibility to ensure cfg is valid using
// cfg.Validate.
func New(cfg config.Config) *Requester {
```

File: requester/requester.go
Comment (GregoryAlbouy) [reply requester/requester.go:78]: Oh nice catch indeed. Will fix it as you suggest, and propose a second PR with mutexes targeting this one so we can compare.

File: requester/requester.go
Comment (moreirathomas) [reply requester/requester.go:34]: But -1 _is_ a valid config isn't it ? Or did I miss a change ?

Thus we can pass -1 in `make`.

Comment (GregoryAlbouy) [reply requester/requester.go:34]: Damn why do I keep forgetting that.

File: requester/report.go
Comment (moreirathomas) [inline requester/report.go:48]: Why not call `cancel` here ? We stopped the process by the context with duration countdown is still going, and `errC` is not closed either.

Maybe we should not handle this here, but I think we should handle it anyway.
So either pass `cancel` to `collect` or read error inside `Run` and cancel before returning.

File: requester/report.go
Comment (moreirathomas) [inline requester/report.go:13]: Why 128 specifically ? Does it get reallocated if it exceeds?

File: requester/report.go
Comment (GregoryAlbouy) [reply requester/report.go:48]: `cancel` is already called in `Run`. It has nothing to do with error handling: we use `cancel` to close it once we're done we it to avoid leaks (similarly to `io.ReadCloser`s `Close`), so the right place to do it is just after calling `Dispatcher.Do(ctx, ...)`, regardless of how we handle the error it returns.

File: requester/report.go
Comment (GregoryAlbouy) [reply requester/report.go:13]: Because why not 😛 The cap algorithm is very simple: it reallocates the double when reached. So for 1000 requests, it'll be reallocated 3 times.
These amount of space are not gonna be an issue, so I think we can go for `1000` by default.

File: requester/report.go
Comment (moreirathomas) [inline requester/report.go:28]: Should not happen but should we noop if `Run` was not called, ie `len(r.records) == 0 && r.numErr == 0` ?

Comment (moreirathomas) [inline requester/report.go:28]: Should not happen but should we noop if `Run` was not called, ie `len(r.records) == 0 && r.numErr == 0` ?

**Review by moreirathomas — APPROVED** [review]
Small question of small incidence. I'll approve right now, but if you think you need to address it, i'll re-approve again :)

File: requester/report.go
Comment (GregoryAlbouy) [reply requester/report.go:28]: I wouldn't bother given it's private + it would return a zero Report anyway (except for the config).
We can consider making it a package func as a foolproof:
```go
func makeReport(cfg config.Config, records []Record, numErr int) Report
```

Comment (moreirathomas) [reply requester/report.go:28]: Let's put that as a "nice to have" and move on then :)

Comment (GregoryAlbouy) [reply requester/report.go:28]: I'm fine making the change if we judge it preferable, it's a matter of seconds

File: dispatcher/dispatcher_test.go
Comment (c1-ra) [inline dispatcher/dispatcher_test.go:79]: [question]
Is `expIterMax` determined empirically too?

File: requester/report.go
Comment (c1-ra) [reply requester/report.go:28]: If it is quick, I think it does not eat bread :bread: to add it. I guess it makes more explicit that we have thought about this possibility and that it will not return a report in this case.

**Review by c1-ra — APPROVED** [review]
:+1:

**Review by moreirathomas — APPROVED** [review]

**Review by c1-ra — APPROVED** [review]
Nice!

**Pull request merged**
